sql/core/src/main/scala/org/apache/spark/sql/execution/SubqueryBroadcastExec.scala

```diff
@@ -75,17 +75,21 @@ case class SubqueryBroadcastExec(
         val beforeCollect = System.nanoTime()
 
         val broadcastRelation = child.executeBroadcast[HashedRelation]().value
-        val (iter, expr) = if (broadcastRelation.isInstanceOf[LongHashedRelation]) {
-          (broadcastRelation.keys(), HashJoin.extractKeyExprAt(buildKeys, index))
-        } else {
-          (broadcastRelation.keys(),
-            BoundReference(index, buildKeys(index).dataType, buildKeys(index).nullable))
+        val rows = broadcastRelation match {
+          case EmptyHashedRelation =>
+            Array.empty[InternalRow]
```

Comment:
Ah, nice catch. We cannot just return an empty row in `EmptyHashedRelation.keys()`?
```
case object EmptyHashedRelation extends HashedRelation {
  ...
  override def keys(): Iterator[InternalRow] = {
    Iterator.empty[InternalRow]
  }
```

Reply:
I'm also surprised it throws an exception instead of returning an empty iterator. I'm not sure if there is any particular reason to do this. But I'll try to update as your suggestion and see if it works, thanks!

Reply:
also cc @cloud-fan and the original author @leanken

Reply:
it's originally throws an exception because it's supposed not to be called. But since you found a exception case here, i think it should return a empty iterator instead.

Reply:
thanks for catching this. ^_^

Reply:
@leanken Thanks for clarification. I'll update